ccontrol-1.0/gui/gccontrol.1
'\" t
.\"     Title: gccontrol
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 2006-08-06
.\"    Manual: [FIXME: manual]
.\"    Source: [FIXME: source]
.\"  Language: English
.\"
.TH "GCCONTROL" "1" "2006\-08\-06" "[FIXME: source]" "[FIXME: manual]"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gccontrol \- GUI for ccontrol
.SH "SYNOPSIS"
.PP
The
\fBgccontrol\fR
command takes no arguments\&.
.SH "DESCRIPTION"
.PP
\fBgccontrol\fR
provides a Graphical User Interface for choosing the configuration file for
\fBccontrol\fR
to use\&. It works by switching the\ \&
\fB$HOME\fR/\&.ccontrol/config
symlink to point to another configuration file in that directory\&.
.SH "SEE ALSO"
.PP
ccontrol(1), ccontrol\-init(1)\&.
.SH "AUTHOR"
.PP
This manual page was written by
Ted
Percival
<ted@midg3t\&.net>
for the Debian system (but may be used by others)\&. Permission is granted to copy, distribute and/or modify this document under the terms of the
GNU
General Public License, Version 2 any later version published by the Free Software Foundation\&.
.PP
On Debian systems, the complete text of the GNU General Public License can be found in /usr/share/common\-licenses/GPL\&.
.SH "COPYRIGHT"
.br
Copyright \(co 2006 Ted Percival
.br


ccontrol-1.0/CHANGES
Changes for 1.0
---------------
Configuration language changes:
	- config file now called ~/.ccontrol/config not default
	- '= off' now replaced with 'disable', can be used in more places
	- 'parallel' keyword gone: use 'no-parallel disable'
	- 'no-parallel' by itself no longer valid, use 'no-parallel = *'
	- 'nice' keyword to change priority
	- 'lock-file' keyword to have some builds not block others.
	- Paths (include, distcc, etc) expand ~ so you can use "~/bin/gcc".
	- Unknown argv[0] is passed through, so you can ccontrol anything.
	

Changes for 0.9
---------------
New features:
	- Gnome panel control-area applet which sets symlink

Other changes:
	- Fix PowerPC (and other?) locking (patch from Tony Breeds)
	- Intercept ccache misses, so we don't have to guess chance of hit
	- Fix error reporting (also patched by Olof Johansson)

Changes for 0.8
---------------
Other changes:
	- fix bug where all complations done at once if distcc failed.
	- invoking ccontrol by itself gives current settings.
	- invoking ccontrol -V/--version gives current version.
	- debian packaging
	- fix bug where ccontrol didn't reread config file when re-symlinked

Changes for 0.7
---------------
Configuration language changes:
	- "add make" to add makefile arguments (eg. for cross-compiling).

Other changes:
	- ccontrol can now be invoked with --section= for special effects.
	- Faster response to config changes if distcc fails.
	- Examples and troubleshooting in documentation

Changes for 0.6
---------------
Configuration language changes:
	- distc++-hosts for c++ compiles (defaults to distcc-hosts)
	- parallel keyword to override no-parallel
	- distcc only used if distcc-hosts/distc++-hosts set

Other changes:
	- toplevel make now shouldn't queue behind other makes.

Changes for 0.5
---------------
Configuration language changes:
	- off keyword for distcc and ccache to turn them off locally

Other changes:
	- ccontrol: "cannot decrement semaphore" fixed: support multiple users.
	- distcc: detect when distcc is failing, don't flood localhost.
	- documentation: fix cpus entry in ccontrol(1)
	- ccontrol-init: human names, fix some failure cases
	- ccache: don't limit distcc so hard when ccache enabled
	- install: create directories if necessary

Changes for 0.4
---------------
IMPORTANT: Configuration language changes:

	- "parallel" renamed to "cpus" (other was confusing)

	- "[global]" must be replaced by "[*]": we now process *every*
	  matching section, in order.

Other changes:
	ccontrol-init: better handling for probing distcc hosts
	configuration: "include" keyword for including files
	documentation: clarifications, hacking advice, license
	distribution: include .hg dir to help users become developers
	users: ccontrol now has a second user (and developer)
	


ccontrol-1.0/.pc/image-in-datadir/gui/gccontrol.py
#!/usr/bin/python
# Simple GTK+ frontend to ccontrol

import gtk
import egg.trayicon
import os
import stat

LIBDIR = "/usr/lib/ccontrol"


def unexpand_user(path):
    home = os.getenv('HOME')
    common = os.path.commonprefix([path, home])
    if len(common) > 0:
        return '~%s' % path[len(common):]
    else:
        return path


class ConfigNotSymlinkException(Exception):
    def __init__(self, file, dir):
        self.file = file
        self.dir = dir


class ConfigManager(object):
    CC_DIR  = os.path.join(os.getenv('HOME'), '.ccontrol')
    CONFIG = os.path.join(CC_DIR, 'config')

    def config_files(self):
        try:
            for file in os.listdir(self.CC_DIR):
                path = os.path.join(self.CC_DIR, file)
                if not os.path.isfile(path) or os.path.islink(path) or \
                        file.startswith('.') or file.endswith('~'):
                    continue
                yield file
        except OSError, e:
            if getattr(e, 'errno', None) == 2:
                return

    def get_config(self):
        if os.path.islink(self.CONFIG):
            return os.readlink(self.CONFIG)
        elif os.path.exists(self.CONFIG):
            raise ConfigNotSymlinkException(self.CONFIG, self.CC_DIR)
        else:
            return None

    def set_config(self, file):
        # Remove file if it already existed
        if os.access(self.CONFIG, os.F_OK):
            os.unlink(self.CONFIG)
        os.symlink(file, self.CONFIG)

class ConfigEditor(gtk.Window):
    class ConfigTab(gtk.Frame):
        def __init__(self, file):
            gtk.Frame.__init__(self)
            self.file = file

        def save(self):
            pass

    def button_response(self, resp):
        if resp == "cancel":
            self.hide_all()
        elif resp == "ok":
            for p in range(0,self.notebook.get_n_pages()):
                self.notebook.get_nth_page(p).save()
            self.hide_all()

    def __init__(self):
        gtk.Window.__init__(self)

        self.set_title("Manage ccontrol profiles")
        self.set_default_size(400, 400)

        vbox = gtk.VBox()

        self.notebook = gtk.Notebook()

        cm = ConfigManager()
        for f in cm.config_files():
            self.notebook.append_page(ConfigEditor.ConfigTab(f), gtk.Label(f))

        vbox.add(self.notebook)

        buttons = gtk.HButtonBox()

        ok_button = gtk.Button(stock='gtk-ok')
        ok_button.connect_object("clicked", self.button_response, "ok")
        buttons.add(ok_button)

        cancel_button = gtk.Button(stock='gtk-cancel')
        cancel_button.connect_object("clicked", self.button_response, "cancel")
        buttons.add(cancel_button)

        vbox.add(buttons)

        self.add(vbox)

class ProfilePopup(gtk.Menu):
    def menuitem_response(self, file):
        try:
            self._cm.set_config(file)
        except Exception, e:
            dialog = gtk.MessageDialog(None, flags=gtk.DIALOG_MODAL,
                    type=gtk.MESSAGE_ERROR, buttons=gtk.BUTTONS_CLOSE)
            dialog.set_markup('<b>An error occurred switching ' \
                    'configuration files!</b>')
            dialog.format_secondary_markup('The error was: <tt>%s</tt>' % e)
            dialog.connect("response", lambda dialog, resp: dialog.destroy())
            dialog.show()
            raise e

    def __init__(self, event):
        gtk.Menu.__init__(self)
        self._cm = ConfigManager()

        try:
            current = self._cm.get_config()
        except ConfigNotSymlinkException, e:
            dialog = gtk.MessageDialog(None, flags=gtk.DIALOG_MODAL,
                    type=gtk.MESSAGE_ERROR,
                    buttons=gtk.BUTTONS_CLOSE)
            dialog.set_markup('<b><tt>%s</tt> is not a symbolic link.</b>' % \
                        unexpand_user(e.file))
            dialog.format_secondary_markup('To use the ccontrol GUI you ' \
                        'must symlink <tt>%s</tt> to a configuration file ' \
                        'in <tt>%s</tt>' % (unexpand_user(e.file),
                            unexpand_user(e.dir)))
            dialog.connect("response", lambda dialog, resp: dialog.destroy())
            dialog.show()
            return

        found = False
        for f in self._cm.config_files():
            found = True
            item = gtk.CheckMenuItem(f)
            item.set_draw_as_radio(True)
            if f == current:
                item.set_active(True)
            self.append(item)
            item.connect_object("activate", self.menuitem_response, f)

        if not found:
            item = gtk.MenuItem('No config files found')
            item.set_property('sensitive', False)
            self.append(item)

        self.show_all()
        self.popup(None, None, None, event.button, event.time)

class TrayIcon(egg.trayicon.TrayIcon):
    def __init__(self):
        egg.trayicon.TrayIcon.__init__(self, 'ccontrol')

        self._gui = ConfigEditor()

        img = gtk.Image()
        img.set_from_file("%s/ccontrol-key.png" % LIBDIR)
        eventbox = gtk.EventBox()
        eventbox.add(img)
        eventbox.set_events(gtk.gdk.BUTTON_PRESS_MASK)
        eventbox.connect_object("button_press_event", self.tray_clicked, self)
        self.add(eventbox)

    def tray_clicked(self, widget, event):
        if event.button == 1:
            ProfilePopup(event)
        elif event.button == 3:
            self._gui.show_all()

def main():
    tray_icon = TrayIcon()
    tray_icon.show_all()
    gtk.main()

main()
